jupyter_server/services/contents/manager.py

```diff
@@ -459,8 +459,12 @@ def copy(self, from_path, to_path=None):
         from_path must be a full path to a file.
         """
         path = from_path.strip("/")
+
         if to_path is not None:
             to_path = to_path.strip("/")
+            dir_exists = self.dir_exists(to_path)
+            if not dir_exists:
```

Comment:
We should probably leverage the `self.dir_exists(to_path)` at L483 below to avoid a redundant dir-exists check.

It's interesting, but the REST API appears to only support "copy" into a directory (typically the same directory as the source), whereas the ContentsManager also supports copy to a file, which is why the tests are failing (due to the first `dir_exists()` call since `to_path` represents a file).

It seems like the case that's missing here (even before this PR) is when `to_path` (being either a directory or file) references a non-existent directory.  The way things stand now, this will result in a 500 error similar to the following (reproduced by changing a character in the parent directory name from `'b'` to `'x'` in the `test_copy` test):
```
tornado.web.HTTPError: HTTP 500: Internal Server Error (Unexpected error while saving file: å x/copy 2.ipynb [Errno 2] No such file or directory: '/private/var/folders/4w/qxgsbhyx1y93qlk9mnzwjryc0000gn/T/pytest-of-kbates/pytest-113/test_copy_jp_contents_manager30/å x/copy 2.ipynb')

jupyter_server/services/contents/filemanager.py:806: HTTPError
```
In order to preserve parity with the handler (where `to_path` can only be a directory and a 404 is returned if it doesn't exist), I think some additional logic is required to check if `to_path` is only a directory name or a combination of directory/file and ensure the respective directory portion exists (throwing 404 if it does not).

(Same would be needed in the async CM as well.)

Reply:
Good eye. Made the modifications and tried to preserve the parity you mentioned with a few test cases to verify.